File: src/main/java/pl/khuzzuk/wfrp/helper/model/rule/Placement.java
package pl.khuzzuk.wfrp.helper.model.rule;

import javax.validation.Constraint;
import javax.validation.ConstraintValidator;
import javax.validation.ConstraintValidatorContext;
import javax.validation.Payload;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import java.util.EnumSet;
import java.util.Set;

public enum Placement {
    BODY,
    HEAD,
    TORSO,
    HAND,
    LEG,
    BELT,
    NECK,
    FINGER,
    SHIELD,
    BOTH_HANDS;

    @Retention(RetentionPolicy.RUNTIME)
    @Target(ElementType.FIELD)
    @Constraint(validatedBy = Validator.class)
    public @interface PlacementValues {
        Placement[] value() default {};
        String message() default "Wrong placement value";
        Class<?>[] groups() default {};
        Class<? extends Payload>[] payload() default {};
    }

    public static class Validator implements ConstraintValidator<PlacementValues, Object> {
        private EnumSet<Placement> values;

        @Override
        public void initialize(PlacementValues constraintAnnotation) {
            Set<Placement> value = Set.of(constraintAnnotation.value());
            values = EnumSet.noneOf(Placement.class);
            values.addAll(value);
        }

        @Override
        public boolean isValid(Object value, ConstraintValidatorContext context) {
            return values.contains(value);
        }
    }
}

File: src/wfrp-helper-web/src/model/crafting/Placement.js
import {State} from "../../state/State";

export const Placement = {
    BODY: 'BODY',
    HEAD: 'HEAD',
    TORSO: 'TORSO',
    HAND: 'HAND',
    LEG: 'LEG',
    BELT: 'BELT',
    NECK: 'NECK',
    FINGER: 'FINGER',
    SHIELD: 'SHIELD',
    BOTH_HANDS: 'BOTH_HANDS',

    allOf() {
        return [
            this.BODY,
            this.HEAD,
            this.TORSO,
            this.HAND,
            this.LEG,
            this.BELT,
            this.NECK,
            this.FINGER,
            this.SHIELD,
            this.BOTH_HANDS,
        ];
    },

    armor() {
        return [
            this.BODY,
            this.HEAD,
            this.TORSO,
            this.HAND,
            this.LEG,
            this.BELT,
            this.SHIELD,
        ];
    },

    weapon() {
        return [
            this.HAND,
            this.BOTH_HANDS,
            this.SHIELD,
        ];
    },

    jewelry() {
        return [
            this.NECK,
            this.FINGER,
        ];
    },
};

export const placementName = 'placement';
export const placementArmorName = 'placementArmor';
export const placementWeaponName = 'placementWeapon';
export const placementJewelryName = 'placementJewelry';

State.data[placementName] = Placement.allOf();
State.data[placementArmorName] = Placement.armor();
State.data[placementWeaponName] = Placement.weapon();
State.data[placementJewelryName] = Placement.jewelry();